FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,50 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let newCache = try ManagedCache.newUniqueInstance(in: context)
+				newCache.timestamp = timestamp
+				newCache.feed = ManagedFeedImage.images(from: feed, in: context, with: newCache)
+
+				try context.save()
+				completion(nil)
+
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.deletePreviousCache(in: context)
+				try context.save()
```

Comment:
In case there is no previously stored cache `ManagedCache.deletePreviousCache` will not change the context, thus there will be no changes in it and will be no need to save it. Otherwise, Core Data may perform unnecessary work.
Since there's no information on whether there was actual deletion or not in this scope then `NSManagedObjectContext.hasChanges` property can be used to make a decision.

Please add `hasChanges` check before saving to avoid unnecessary work.
```suggestion
				if context.hasChanges { 
      					try context.save()
      				}
```